### ISSUE TITLE
Questions Integration Testing Hasura. 

### ISSUE THREAD

**Issue by williscool** [hasura/graphql-engine#5561]
Tl:dr

1. What are the best practices around integration testing hasura?
2. Is there currently away to support db isolation via templates for test i.e. changing the db url?
3. follow q2 here interesting solution https://github.com/allaboutapps/integresql#usage  is it possible to setup hasura to be compatible?

---

I'm using jest to integration test a hasura application im working on.

Given that jest runs tests in parallel by default and my tests assume a clean database you can imagine I quickly ran into intermittent test breaks with database state being wrong for a certain test.

I've mostly fixed the issues via `jest --runInBand` to serialize the tests but obviously it makes the suite a lot slower

here is the trashy db rollback code I've been using

``` javascript
/**
 * name says it all
 */
export async function nukeDB() {
  const query = gql`
    mutation nukeDB {
      delete_users(where: {}) {
        affected_rows
      }
      delete_item(where: {}) {
        affected_rows
      }
      delete_notifications(where: {}) {
        affected_rows
      }
    }
  `;

  
  return await hasuraGqlClient.request(query);;
}
```

My questions are in the tl:dr so I'll add more background here

while researching this I stumbled upon this early promising project

https://github.com/allaboutapps/integresql

https://github.com/allaboutapps/integresql#final-approach-integresql

they have a service you can run in a docker container that will allow you to create dbs programatically based on templates.

If I could figure out how to have each test isolate the db hasura uses for them (like via a header)

I think this would be really powerful for testing.

Thanks for the great project!

**Comment by tirumaraiselvan** [comment]
Hi there! Apologies for the late response and great question. While the best setup for integration tests will depend on your environment and your requirements, I can suggest few common methods:

If you need a fresh DB instance per test, then you may create a different postgres db for each of your parallel test worker (`postgresql://user:pass@localhost:5432/test1`, `postgresql://user:pass@localhost:5432/test2`, `postgresql://user:pass@localhost:5432/test3`) . You will need to apply the hasura metadata for each of the DBs as well.

You can also test your schema/metadata changes for regressions via [Hasura Regression Testing](https://hasura.io/docs/1.0/graphql/cloud/regression-tests.html).

This blogpost gives a high-level overview of the general workflow: https://hasura.io/blog/moving-from-local-development-staging-production-with-hasura/#moving-to-staging
